nixos/tor: fixes for settings.HidServAuth

tomfitzhenry opened the pull request:
###### Motivation for this change
fixes https://github.com/NixOS/nixpkgs/issues/122429

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

tomfitzhenry committed:
nixos/tor: fix HidServAuth

tomfitzhenry committed:
nixos/tor: add an example for services.tor.settings.HidServAuth

tomfitzhenry commented on nixos/modules/services/security/tor.nix:
```diff
@@ -170,7 +170,7 @@ let
     else if k == "ServerTransportPlugin" then
       optionalString (v.transports != []) "${concatStringsSep "," v.transports} exec ${v.exec}"
     else if k == "HidServAuth" then
-      concatMapStringsSep "\n${k} " (settings: settings.onion + " " settings.auth) v
```
ju1m@, PTAL.

dotlambda commented on nixos/modules/services/security/tor.nix:
```diff
@@ -715,7 +715,7 @@ in
                 options = {
                   onion = mkOption {
-                    type = strMatching "[a-z2-7]{16}(\\.onion)?";
+                    type = strMatching "[a-z2-7]{16,56}(\\.onion)?";
```
```suggestion
                    type = strMatching "([a-z2-7]{16}|[a-z2-7]{56})(\\.onion)?";
```

dotlambda replied:
Btw, do you know if `.onion` is actually optional? That's not mentioned on https://2019.www.torproject.org/docs/tor-manual.html.en, but that documentation is outdated anyway and doesn't even mention v3 addresses.

dotlambda reviewed (commented):
Shouldn't we use `listsAsDuplicateKeys = true`?

dotlambda commented:
Can we add a test case to `nixos/tests/tor.nix`?

ju1m reviewed (changes requested):
Support for v3 addresses in `HidServAuth` must not be added, besides LGTM.

tomfitzhenry commented:
> Can we add a test case to nixos/tests/tor.nix?

I just tried this for 30 min, and it's surprisingly fiddly to test a hidden services. It requires setting up a directory authority, a relay, a server, a client, and coordinating various keys in various formats between the different actors.

tomfitzhenry committed:
nixos/tor: fix HidServAuth validation to require ".onion"

Per https://manpages.debian.org/testing/tor/torrc.5.en.html :

> Valid onion addresses contain 16 characters in a-z2-7 plus ".onion"

tomfitzhenry commented:
> @tomfitzhenry you may still want to use removeSuffix ".onion" v.onion + ".onion" or remove the ? in (\\.onion)?

Done. I opted for the latter, to avoid the maintenance burden of allowing different input types. This is consistent with torrc. Removing support for hostnames without ".onion" won't break anyone, since this functionality has never worked anyway.
